### PR TITLE
UAF-2554 - MOD Global Transaction Edit (GTE)

### PR DESCRIPTION
**Pull request by sskinner**
Adding GTE Mod implementation to establish equivalency with KFS3. Additions include all Java, Spring, OJB, data dictionary, JSP, and the struts config files. This represents the entire GTE mod, including added GTE validation in the GEC, PDP, and LD modules. Lastly, this code also introduces added validation for batch's GL collector and LD scrubber. Note, our DB upgrade scripts preserve the one needed param and two tables that this mod is dependent upon, so no liquibase scripts were necessary in this mod's PR.

### CODE REVIEW

File: kfs-core/src/main/java/edu/arizona/kfs/fp/businessobject/ErrorCertification.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public String getErrorCorrectionReason() {
 	public void setErrorCorrectionReason(String errorCorrectionReason) {
 		this.errorCorrectionReason = errorCorrectionReason;
 	}
-	
+
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/businessobject/ErrorCertification.java:54]: What is the reason for this change? Formatting?

Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/businessobject/ErrorCertification.java:54]: Yes, I tried as hard as possible to avoid it, but when having to deal with the multiple merge conflicts from having pulled in effectively ~20 other commits from the ua-development branch, a couple formatting changes did get through. I spent an entire day trying to iron it all out, and these three squashed commits were the best I could manage. While on that subject, I'd like to discuss standardizing how we format new files that are pulled in, so that we can avoid this entirely in the future.

File: kfs-core/src/main/java/edu/arizona/kfs/fp/document/validation/impl/NonCheckDisbursementRequiredReferenceFieldValidation.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            clazz = document.getTargetAccountingLineClass();
+        }
+
+        boe = getDataDictionaryService().getDataDictionary().getBusinessObjectEntry(clazz.getName());
+
+        if (StringUtils.isEmpty(getAccountingLineForValidation().getReferenceNumber())) {
+            putRequiredPropertyError(boe, REFERENCE_NUMBER);
+            valid = false;
+        }
+
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/document/validation/impl/NonCheckDisbursementRequiredReferenceFieldValidation.java]: Why not execute validate on super and then add the GTE specific validation? Seems like lines 43-60 are just a duplication of the overridden super method code.

Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/document/validation/impl/NonCheckDisbursementRequiredReferenceFieldValidation.java]: Why not indeed -- change made and pushed.

File: kfs-core/src/main/java/edu/arizona/kfs/gl/document/validation/GlobalTransactionEditDetailRule.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
+/*
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/document/validation/GlobalTransactionEditDetailRule.java]: This is the old license. My position on licenses in the edu.arizona code files is either no license (preferred) or the new license.

Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/document/validation/GlobalTransactionEditDetailRule.java]: Removed and pushed.

File: kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        if (valid) {
+            valid = checkForMixedBalanceTypes(batch, messageMap);
+        }
+
+        if (valid) {
+            valid = checkDetailKeys(batch, messageMap);
+        }
+
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java]: Why not execute performValidation on super and then add the GTE specific validation? Seems like lines 45-63 are just a duplication of the overridden super method code.

Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java]: In fact it can be done, but my intent was to do a straight port as requested. Since what I found in the old baseline is that most of the code was implemented in the Kuali packages, I aimed to extract the minimal amount needed to override into the Arizona packages. In this instance it's pretty clear that we could call super easily enough, but further down in other examples it gets way more fuzzy. So in the interest of trying to stick to an apples-to-apples port, keeping the process expedient, and ensuring logic intent was preserved, the compromise was pulling the whole method up. If this is not acceptable, we should stop the code review now, because this will introduce significantly more left to do, as this is the strategy I applied to the entire port.

Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java]: Not calling super when we can do so introduces duplicate code into our application. This is not good coding practice, for obvious reasons. If you copy an entire method over to edu.arizona, add the new required functionality and then make changes to fix a bug, bad coding or bad design, then I am OK with it.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/AuxiliaryVoucherValidation.xml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+<beans xmlns="http://www.springframework.org/schema/beans"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.springframework.org/schema/beans
+        http://www.springframework.org/schema/beans/spring-beans-2.0.xsd">
+
+  <bean id="AuxiliaryVoucher-accountingLineValuesAllowedValidation" class="edu.arizona.kfs.sys.document.validation.impl.AccountingLineValuesAllowedValidationHutch" abstract="true">
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/AuxiliaryVoucherValidation.xml]: The KFS 6.0 document validation configuration file AuxiliaryVoucherValidation.xml does not contain a bean AuxiliaryVoucher-accountingLineValuesAllowedValidation. It uses bean AccountingDocument-defaultAccountingLineValuesAllowedValidation from file FinancialSystemValidators.xml. I don't believe this override is needed.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+<beans xmlns="http://www.springframework.org/schema/beans"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.springframework.org/schema/beans
+        http://www.springframework.org/schema/beans/spring-beans-2.0.xsd">
+
+  <!-- Had to include a bean in orig. parent bean because it had no list to merge with -->
+  <bean id="DistributionOfIncomeAndExpense-addAccountingLineValidation" parent="DistributionOfIncomeAndExpense-addAccountingLineValidation-parentBean" scope="prototype">
+    <property name="validations">
+      <list>
+        <bean parent="AccountingDocument-AddAccountingLine-DefaultValidation" scope="prototype"/>
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Should be: bean parent="AccountingDocument-AddCapitalAccountingLines-DefaultValidation" scope="prototype"

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        </bean>
+      </list>
+    </property>
+  </bean>
+
+  <!-- Had to include a bean in orig. parent bean because it had no list to merge with -->
+  <bean id="DistributionOfIncomeAndExpense-updateAccountingLineValidation" parent="DistributionOfIncomeAndExpense-updateAccountingLineValidation-parentBean" scope="prototype">
+    <property name="validations">
+      <list>
+        <bean parent="AccountingDocument-UpdateAccountingLine-DefaultValidation" scope="prototype"/>
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Should be: bean parent="AccountingDocument-UpdateCapitalAccountingLines-DefaultValidation" scope="prototype"

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml (hunk 6 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Interesting, since this is exactly how it is in KFS3:
<bean parent="AccountingDocument-AddAccountingLine-DefaultValidation" scope="prototype" />

And in fact, the KFS3 kuali package parent is also:
<bean parent="AccountingDocument-AddAccountingLine-DefaultValidation" scope="prototype" />

So to me, it appears the KFS6 parent may be incorrect. what do you make of it? Makes me nervous changing from how it has been and is out in PRD now, and especially now with the question of if v6 kuali package parent is the mistake.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml (hunk 7 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Same thing as for the add validation, where this differs than in KFS3, and the kuali parent matches what I have here. So is it a change between versions, or is the KFS6 kuali parent incorrect?

Edit: maybe my brain is getting tired, and I should revisit this in the morning.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml (hunk 6 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Ok, after more sleep, and actually tracing this, I see that this doesn't hurt anything when changed. What happens is the regular validation still happens, but if it's a Capital Accounting Line that is involved, slightly different logic also gets applied to determine line accessibility. This is brand new in v6, hence why it didn't carry over from v3. I will change this here and for the update validation.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml (hunk 7 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml]: Please see my comment for AccountingDocument-AddCapitalAccountingLines; I am changing this per recommendation too.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/AuxiliaryVoucherValidation.xml (hunk 5 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/AuxiliaryVoucherValidation.xml]: I think you're right. In  v3, the behavior is really odd given that in some spots spring changes are directly in the kuali package, while others are properly overridden in the az package. When I load the v3 spring context in Intellij, and follow references for the above, I see a kuali spring file that points to the az bean we see here -- this shouldn't ever happen, and shows how v3 has some undesirable behavior. I'll delete this file, since this is already provided in the system validators as you observed (and as I verified in the spring explorer).

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/GeneralErrorCorrectionValidation.xml [hunk 8]
```diff
@@ -1,25 +1,18 @@
...
 				<bean parent="GeneralErrorCorrection-errorCertificationValidation" scope="prototype"/>
 				<bean parent="GeneralErrorCorrection-errorCertificationTabValidation" scope="prototype"/>
 			</list>
 		</property>
 	</bean>
-    
+
     <bean id="GeneralErrorCorrection-addAccountingLineValidation" parent="GeneralErrorCorrection-addAccountingLineValidation-parentBean" scope="prototype">
         <property name="validations">
             <list>
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/GeneralErrorCorrectionValidation.xml:18]: Missing from list: bean parent="AccountingDocument-AddCapitalAccountingLines-DefaultValidation"

Also in GeneralErrorCorrection-updateAccountingLineValidation first bean in list should be bean parent="AccountingDocument-UpdateCapitalAccountingLines-DefaultValidation" scope="prototype" not bean parent="AccountingDocument-UpdateAccountingLine-DefaultValidation" scope="prototype".

There was no way to add a line comment at that point.

Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/GeneralErrorCorrectionValidation.xml:18]: Both have been corrected (again, missed due to being new in v6). Also, I am saving up all of these changes to push under one commit, so as to keep the git log a bit cleaner.

File: kfs-core/src/main/resources/edu/arizona/kfs/gl/gl-resources.properties [hunk 9]
```diff
@@ -0,0 +1 @@
+error.gl.GlobalTransactionEdit=Accounting line transaction is not allowed because it violates the following Global Transaction Edit validation rule: Reference Origin Code={0}, Fund Group={1}, Sub-Fund Group={2}, Document Type={3}, Object Type={4}, Object Sub-Type={5}.
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/gl/gl-resources.properties:1]: Where is the text for the other three GTE error messages (see KFSKeyConstants.java)?

File: kfs-core/src/main/resources/institutional-config.properties [hunk 10]
```diff
@@ -9,7 +9,7 @@ ecustoms.last.run.date.default=2013-01-01 00:00:00
...
 rice.struts.message.resources=org.kuali.rice.krad.ApplicationResources,org.kuali.rice.krad.KRAD-ApplicationResources,org.kuali.rice.kew.ApplicationResources,org.kuali.rice.ksb.ApplicationResources,org.kuali.rice.kim.ApplicationResources,org.kuali.rice.krms.ApplicationResources,org.kuali.rice.core.web.cache.CacheApplicationResources,org.kuali.kfs.coa.coa-resources,org.kuali.kfs.fp.fp-resources,org.kuali.kfs.gl.gl-resources,org.kuali.kfs.pdp.pdp-resources,org.kuali.kfs.sys.sys-resources,org.kuali.kfs.vnd.vnd-resources,org.kuali.kfs.module.ar.ar-resources,org.kuali.kfs.module.bc.bc-resources,org.kuali.kfs.module.cab.cab-resources,org.kuali.kfs.module.cam.cam-resources,org.kuali.kfs.module.cg.cg-resources,org.kuali.kfs.module.ec.ec-resources,org.kuali.kfs.module.external.kc.kc-resources,org.kuali.kfs.module.ld.ld-resources,org.kuali.kfs.module.tem.tem-resources,org.kuali.kfs.module.purap.purap-resources,edu.arizona.kfs.module.purap.purap-resources,edu.arizona.kfs.coa.coa-resources,edu.arizona.kfs.fp.fp-resources
```
Comment (ghost) [inline kfs-core/src/main/resources/institutional-config.properties]: Need to add ",edu.arizona.kfs.gl.gl-resources" to the end of property rice.struts.message.resources.

Comment (sskinner) [inline kfs-core/src/main/resources/institutional-config.properties]: Well if that isn't gotcha; I wasn't aware of this second one at all, and the name of this key is really deceptive with what the final constant that picks it up is named. Corrected.

File: kfs-core/src/main/resources/edu/arizona/kfs/gl/gl-resources.properties (hunk 9 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/gl/gl-resources.properties:1]: They are keys used to place errors in the error message map by various validations, services, or helpers to indicate where a validation had failed, thus we don't see correlating error message values for them in this file.

File: kfs-core/src/main/resources/edu/arizona/kfs/pdp/spring-pdp.xml [hunk 11]
```diff
@@ -18,6 +11,32 @@
...
+        <property name="subObjectCodeService" ref="subObjectCodeService"/>
+        <property name="businessObjectService" ref="businessObjectService"/>
+        <property name="bankService" ref="bankService"/>
+        <property name="originationCodeService" ref="originationCodeService"/>
+        <property name="documentTypeService" ref="rice.kew.documentTypeService"/>
+        <property name="dataDictionaryService" ref="dataDictionaryService"/>
+        <property name="globalTransactionEditService" ref="globalTransactionEditService" />
+    </bean>
+
+    <bean id="globalTransactionEditService" parent="globalTransactionEditService-parentBean" />
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/pdp/spring-pdp.xml]: globalTransactionEditService is wired/defined in spring-gl.xml and both gl and pdp are in kfs-core, so no need to define globalTransactionEditService again here in spring-pdp.xml.

Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/pdp/spring-pdp.xml]: Good point, at the time I remember thinking it's good to keep the modules disjoint, forgetting I was working under core regardless. Changed.

File: kfs-core/src/main/resources/edu/arizona/kfs/sys/document/validation/configuration/FinancialSystemValidators.xml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        <value>LaborJournalVoucherDocument</value>
+        <value>PaymentRequestDocument</value>
+        <value>PreEncumbranceDocument</value>
+        <value>ProcurementCardDocument</value>
+        <value>PurchaseOrderDocument</value>
+        <value>RequisitionDocument</value>
+        <value>SalaryExpenseTransferDocument</value>
+        <value>ServiceBillingDocument</value>
+        <value>UaRequisitionDocument</value>
+        <value>UaDisbursementVoucherDocument</value>
```
Comment (ghost) [inline kfs-core/src/main/resources/edu/arizona/kfs/sys/document/validation/configuration/FinancialSystemValidators.xml]: UaRequisitionDocument and UaDisbursementVoucherDocument do not exist in our KFS 6.0 code. These documents can be removed from the list.

Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/sys/document/validation/configuration/FinancialSystemValidators.xml]: Removed.

File: kfs-core/src/main/java/edu/arizona/kfs/sys/document/validation/impl/AccountingLineGlobalTransactionValidation.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    public void setAccountingLineForValidation(AccountingLine accountingLineForValidation) {
+        this.accountingLineForValidation = accountingLineForValidation;
+    }
+
+    /**
+     * Sets the parameterService attribute value.
+     *
+     * @param parameterService The parameterService to set.
+     */
+    public void setParameterService(ParameterService parameterService) {
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/sys/document/validation/impl/AccountingLineGlobalTransactionValidation.java]: I am not seeing parameter service used anywhere in this class. This class also has the old license.

**Comment by ghost** [comment]
I believe we are missing some Spring wiring for the validators. Please add:

edu.arizona spring-fp.xml (after line 34):
import resource="document/validation/configuration/FinancialProcessingValidators.xml" 
import resource="document/validation/configuration/CreditCardReceiptValidation.xml" 
import resource="document/validation/configuration/DistributionOfIncomeAndExpenseValidation.xml"
import resource="document/validation/configuration/PreEncumbranceValidation.xml"
import resource="document/validation/configuration/ProcurementCardValidation.xml"

edu.arizona spring-sys.xml (after line 28):
import resource="document/validation/configuration/FinancialSystemValidators.xml"

edu.arizona spring-purap.xml (after line 27):
import resource="document/validation/configuration/RequisitionValidation.xml"

File: kfs-ld/src/main/java/edu/arizona/kfs/module/ld/batch/service/impl/SalaryTransferPeriodValidationServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @see SalaryTransferPeriodValidationService
+ */
+@Transactional
+public class SalaryTransferPeriodValidationServiceImpl extends org.kuali.kfs.module.ld.document.service.impl.SalaryTransferPeriodValidationServiceImpl {
```
Comment (ghost) [inline kfs-ld/src/main/java/edu/arizona/kfs/module/ld/batch/service/impl/SalaryTransferPeriodValidationServiceImpl.java]: Just noticed this class was moved from kfs/module/ld/document/service/.impl/ to kfs/module/ld/batch/service/impl/. By mistake or is there a reason?

File: kfs-ld/src/main/resources/edu/arizona/kfs/module/ld/spring-ld.xml [hunk 15]
```diff
@@ -24,6 +19,39 @@
...
+    <bean id="salaryTransferPeriodValidationService" parent="salaryTransferPeriodValidationService-parentBean" />
+    <bean id="salaryTransferPeriodValidationService-parentBean" class="edu.arizona.kfs.module.ld.batch.service.impl.SalaryTransferPeriodValidationServiceImpl" abstract="true">
+        <property name="effortCertificationService" ref="effortCertificationModuleService"/>
+        <property name="documentService" ref="documentService"/>
+        <property name="noteService" ref="noteService"/>
+        <property name="configurationService" ref="kualiConfigurationService"/>
+        <property name="globalTransactionEditService" ref="globalTransactionEditService"/>
+    </bean>
+
+    <bean id="globalTransactionEditService" parent="globalTransactionEditService-parentBean" />
```
Comment (ghost) [inline kfs-ld/src/main/resources/edu/arizona/kfs/module/ld/spring-ld.xml]: This is not really needed here as the optional modules (ld being one) can safely reference anything defined in kfs-core, in this case, in spring-gl.xml.

**Comment by ghost** [comment]
Initial review complete. Please see my comments.

File: kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/GeneralErrorCorrectionValidation.xml (hunk 8 above)
Comment (sskinner) [inline kfs-core/src/main/resources/edu/arizona/kfs/fp/document/validation/configuration/GeneralErrorCorrectionValidation.xml:18]: I just checked up on this again, and it has been corrected, but the comment here persists since the lines the changes were made at, are not the lines this comment was made on.

File: kfs-core/src/main/java/edu/arizona/kfs/sys/document/validation/impl/AccountingLineGlobalTransactionValidation.java (hunk 13 above)
Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/sys/document/validation/impl/AccountingLineGlobalTransactionValidation.java]: Copied directly from v3, but removed the method anyway. Also deleted old license.

**Comment by sskinner** [comment]
Added validator entries to spring wiring as requested:
edu.arizona spring-fp.xml (after line 34, x5)
edu.arizona spring-sys.xml (after line 28, x1)
edu.arizona spring-purap.xml (after line 27, x1)

The github comment system doesn't allow me to reply inline since that comment wasn't made directly on a file.

File: kfs-ld/src/main/java/edu/arizona/kfs/module/ld/batch/service/impl/SalaryTransferPeriodValidationServiceImpl.java (hunk 14 above)
Comment (sskinner) [inline kfs-ld/src/main/java/edu/arizona/kfs/module/ld/batch/service/impl/SalaryTransferPeriodValidationServiceImpl.java]: The way the directory structure displayed in the IDE got me, I saw "blah.service.impl" in the collapsed structure, and missed that the prefix was "batch", not "document". Corrected.

File: kfs-ld/src/main/resources/edu/arizona/kfs/module/ld/spring-ld.xml (hunk 15 above)
Comment (sskinner) [inline kfs-ld/src/main/resources/edu/arizona/kfs/module/ld/spring-ld.xml]: Sure, the service being a singleton will point to the same thing, I just figured it might be a convenience to have it be explicitly identified. Removed anyway, which will be another reason for devs to correctly configure their IDE for spring contexts to follow references, which we should all be doing anyways.

**Comment by sskinner** [comment]
Hello @Perrinek, I believe that with my latest push, this is now ready again for your consideration. If a comment above persists, it's either due to the comment being made on the file where the suggestion didn't coincide with the actual change (x2), I am providing an answer to where only clarification was requested (x2), or we had discussed and arrived at an understanding via email (x1). All other suggestions have been made and pushed.

File: kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+import org.kuali.kfs.gl.batch.CollectorStep;
+import org.kuali.kfs.sys.KFSConstants;
+import org.kuali.rice.krad.util.MessageMap;
+
+/**
+ * The base implementation of CollectorHelperService
+ *
+ * @see org.kuali.kfs.gl.batch.service.CollectorService
+ */
+public class CollectorHelperServiceImpl extends org.kuali.kfs.gl.batch.service.impl.CollectorHelperServiceImpl {
```
Comment (ghost) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java]: Just noticed that this class was also moved. In this case, from kfs/gl/batch/service/impl/ to kfs/gl/service/impl/.

**Comment by ghost** [comment]
Second review complete. Please see my single comment.

File: kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java (hunk 16 above)
Comment (sskinner) [inline kfs-core/src/main/java/edu/arizona/kfs/gl/service/impl/CollectorHelperServiceImpl.java]: Ha! This time from the opposite direction! I think the same tricky type of collapsed package illusion got me; I need to get my prescription glasses renewed, it's been too long apparently. I'll move this now and ping you on hipchat once it's pushed.

**Comment by sskinner** [comment]
Hi @Perrinek, change made. built, and smoke-tested; ready for review at your convenience.

**Comment by ghost** [comment]
Looks good!
